zhmc_prometheus_exporter/zhmc_prometheus_exporter.py

```diff
@@ -2187,18 +2199,85 @@ def main():
                  f"Verbosity level: {VERBOSE_LEVEL}")
 
-        hmccreds_filename = args.c
+        config_filename = args.c
         logprint(logging.INFO, PRINT_V,
-                 f"Parsing HMC credentials file: {hmccreds_filename}")
-        yaml_creds_content = parse_yaml_file(
-            hmccreds_filename, 'HMC credentials file', 'hmccreds_schema.yaml')
-        # metrics is required in the metrics schema:
-        yaml_creds = yaml_creds_content["metrics"]
+                 f"Parsing exporter config file: {config_filename}")
+        config_dict = parse_yaml_file(
+            config_filename, 'exporter config file', 'config_schema.yaml')
+
+        if 'metrics' not in config_dict and 'hmc' not in config_dict:
+            new_exc = ImproperExit(
+                "The exporter config file must specify either the new 'hmc' "
+                "item or the old 'metrics' item, but it specifies none.")
+            new_exc.__cause__ = None  # pylint: disable=invalid-name
+            raise new_exc
+
+        if 'metrics' in config_dict and 'hmc' in config_dict:
+            new_exc = ImproperExit(
+                "The exporter config file must specify either the new 'hmc' "
+                "item or the old 'metrics' item, but it specifies both.")
+            new_exc.__cause__ = None  # pylint: disable=invalid-name
+            raise new_exc
+
+        if 'version' in config_dict and config_dict['version'] != 2:
+            new_exc = ImproperExit(
+                "The exporter config file must have the version 2 format, "
+                "but it specifies the version {} format.".
+                format(config_dict['version']))
+            new_exc.__cause__ = None  # pylint: disable=invalid-name
+            raise new_exc
+
+        if 'metrics' in config_dict and 'hmc' not in config_dict:
+            logprint(logging.WARNING, PRINT_ALWAYS,
+                     "DEPRECATED: The exporter config file has the old version "
+                     "1 format. Please migrate to the new version 2 format "
+                     "(use --help-config). The version 1 format is deprecated "
+                     "and support for it will be removed in a future release.")
+
+            # Convert old format to new format
+            config_dict['version'] = 2
+            old_creds = config_dict['metrics']
+            config_dict['hmc'] = {
+                'host': old_creds['hmc'],
+                'userid': old_creds['userid'],
+                'password': old_creds['password'],
+                'verify_cert': old_creds.get('verify_cert', True),
+            }
+            del config_dict['metrics']
+
+        if 'metric_groups' not in config_dict:
+            logprint(logging.INFO, PRINT_ALWAYS,
+                     "The exporter config file has no 'metric_groups' item. "
+                     "Exporting all metric groups.")
+            config_dict['metric_groups'] = {
+                'cpc-usage-overview': {'export': True},
```

Comment:
minor: instead of hardcoding like this, you could have a list of all the keys, and then just iterate that list and add the corresponding export:True.

The part I definitely dislike about this mass assignment: your keys aren't sorted. I would at least have them alphabetically, so it becomes easier in the future when you want to add/remove keys.

Reply:
The order of metric groups is consistent with the example config file, the help-config, and the metric groups in the metric definition file. I find the maintenance easier if these orders are the same.

Also, the metric groups are not a flat list, but are grouped into classic, DPM, and any mode, so if a metric group gets added to one of these groups it gets added to the end of that group and having the same order makes that easier.

On maintaining a list: That list would basically look like the code here, because it would also contain the export boolean because I'd like to be able to change the default in the future.

Reply:
Having a list to iterate and create the values, instead of hardcoding the value generation repeatedly ... would still allow you to write down that list in whatever order you want it to have. So "order" and "using a list" aren't tied together.